tests/utils/zephyr/test_node_edge.py

```diff
@@ -0,0 +1,366 @@
+        lu_qps = ZPlaneShift(-1, -1)
+        right_down_pcs = [ZNode(xy, ZShape(m=m)) for xy, m in self.right_down_xyms]
+        rd_qps = ZPlaneShift(1, 1)
+        midgrid_pcs = [ZNode(xy, ZShape(m=m)) for xy, m in self.midgrid_xyms]
+        for pc in midgrid_pcs:
+            for s1, s2 in product((-2, 2), (-2, 2)):
+                pc + ZPlaneShift(s1, s2)
+        for pc in left_up_pcs:
+            with self.assertRaises(ValueError):
+                pc + lu_qps
+        for pc in right_down_pcs:
+            with self.assertRaises(ValueError):
+                pc + rd_qps
+
+    def test_add_sub(self) -> None:
+        midgrid_pcs = [ZNode(xy, ZShape(m=m)) for xy, m in self.midgrid_xyms]
+        right_down_pcs = [ZNode(xy, ZShape(m=m)) for xy, m in self.right_down_xyms]
+        left_up_pcs = [ZNode(xy, ZShape(m=m)) for xy, m in self.left_up_xyms]
+
+        for pc in midgrid_pcs + right_down_pcs + left_up_pcs:
+            self.assertEqual(pc + ZPlaneShift(0, 0), pc)
+
+    def test_neighbors_generator_runs(self) -> None:
+        for _ in ZNode((1, 12, 4), ZShape(t=6)).neighbors_generator():
+            _
+
+    def test_zcoord(self) -> None:
+        ZNode((11, 12, 4), ZShape(t=6)).zcoord == ZephyrCoord(1, 0, 4, 0, 2)
+        ZNode((1, 0)).zcoord == ZephyrCoord(1, 0, None, 0, 0)
+        ZNode((0, 1)).zcoord == ZephyrCoord(0, 0, None, 0, 0)
+
+    def test_direction_node_kind(self) -> None:
+        for u in self.u_vals:
+            for j in self.j_vals:
+                for m in self.m_vals:
+                    w_vals = range(2 * m + 1)
+                    z_vals = range(m)
+                    for w in w_vals:
+                        for z in z_vals:
+                            for t in self.t_vals:
+                                k_vals = range(t)
+                                for k in k_vals:
+                                    zn = ZNode(coord=(u, w, k, j, z), shape=(m, t))
+                                    self.assertEqual(zn.direction, u)
+                                    if u == 0:
+                                        self.assertTrue(zn.is_vertical())
+                                        self.assertEqual(zn.node_kind, NodeKind.VERTICAL)
+                                    else:
+                                        self.assertTrue(zn.is_horizontal())
+                                        self.assertEqual(zn.node_kind, NodeKind.HORIZONTAL)
+
+    def test_neighbor_kind(self) -> None:
+        zn = ZNode((0, 1))
+        self.assertTrue(zn.neighbor_kind(ZNode((1, 0))) is EdgeKind.INTERNAL)
+        self.assertTrue(zn.neighbor_kind(ZNode((1, 2))) is EdgeKind.INTERNAL)
+        self.assertTrue(zn.neighbor_kind(ZNode((0, 3))) is EdgeKind.ODD)
+        self.assertTrue(zn.neighbor_kind(ZNode((0, 5))) is EdgeKind.EXTERNAL)
+        self.assertTrue(zn.neighbor_kind(ZNode((0, 7))) is None)
+        self.assertTrue(zn.neighbor_kind(ZNode((1, 6))) is None)
```

Comment:
This could definitely be [parameterized](https://github.com/wolever/parameterized) instead, since they're covering different cases.

Reply:
Also, you can use a specialized assert method, [`assertIs`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertIs), for cleaner error messages.